### PR TITLE
Refactor processing

### PR DESCRIPTION
**Pull request by jeqo**
Several changes has been applied to reduce processing overhead:
- A cache of `service:span` names has been added to the `SpanConsumer` so `SpanStore` doesn't need to process spans to collect this map. Only changes to the cache are propagated downstream.
- `SpanConsumer` return to the approach of publishing only raw `spans` downstream.
- Aggregations (e.g. trace aggregation, service names aggregation, dependencies aggregation) has been moved out of storage, so they can be scaled independently. Aggregation becomes a new mode, enabled by default.
- Span Index (Lucene based) has been refactored and abstracted by a `SpanIndexService` that is used by many parties (e.g. `TraceStoreStream` and `SpanStore`). By doing this a stream process consuming raw spans has been removed from the picture, making queries more consistent (remove dependency on two stores) and lower performance overhead.
- Trace aggregation is using Suppression API on Kafka Streams to reduce overhead by publishing only finished traces: Traces that haven't received a span by a defined period of time. Even though this is not a strict way of defining a trace, it give us enough value to define a complete trace and proceed with dependency aggregation.
- Dependencies as well are stored by date instead of link id (parent:child), and we can have a time representation of how graph grows.

Current dataflow:

- Spans collected -> `SpanConsumer`
- `SpanConsumer` -> topic: `SpanServices`
- `SpanConsumer` -> topic: `Spans`
- topic: `Spans` -> `SpanStore` -> stores: Span Index, Trace Store
- topic: `Spans` -> `TraceRetentionStream` -> topic: `Spans`
- topic: `SpanServices` -> `ServiceAggregationStream` -> topic: `Services`
- topic: `Services` -> `SpanStore` -> stores: Service Store
- topic: `Spans` -> `TraceAggregation` -> topic: `Traces`
- topic: `Traces` -> `DependencyAggregation` -> topic: `TraceDependency`, `Dependencies`
- topic: `Dependencies` -> `SpanStore` -> stores: Dependency Store

### CODE REVIEW

File: autoconfigure/src/main/resources/zipkin-server-kafkastore.yml [hunk 1]
```diff
@@ -8,16 +8,19 @@ zipkin:
       compression-type: ${KAFKA_STORE_COMPRESSION_TYPE:NONE}
       retention-scan-frequency: ${KAFKA_STORE_RETENTION_SCAN_FREQUENCY:86400000}
       retention-max-age: ${KAFKA_STORE_RETENTION_MAX_AGE:604800000}
-      trace-inactivity-gap: ${KAFKA_STORE_TRACE_INACTIVITY_GAP:300000}
-      spans-topic: ${KAFKA_STORE_SPANS_TOPIC:zipkin-spans}
+      trace-inactivity-gap: ${KAFKA_STORE_TRACE_INACTIVITY_GAP:60000}
+      spans-topic: ${KAFKA_STORE_SPANS_TOPIC:zipkin-spans-v1}
```
Comment (adriancole) [inline autoconfigure/src/main/resources/zipkin-server-kafkastore.yml]: what does the v1 mean here?

File: README.md [hunk 2]
```diff
@@ -7,27 +7,34 @@ Kafka-based storage for Zipkin.
...
-------------------------------|------------------------------------|----|--------------------
-                              |                                    |    |
-kafka                         +-->(  raw-spans  )--// potential //-+    +-->( traces )
-                              |                    // enriching // |    |
-                              +-->( light-spans )--// sampling  //-+    +-->( dependencies )
-
---------------------------------------------------------------------------------------------
+                    +----------------------------*zipkin*-------------------------------------------
+                    |                                                           +-->( service:span )
+                    |                                                           +-->( span-index   )
```
Comment (adriancole) [inline README.md]: the proper term for a span index is "spandex"

File: README.md [hunk 3]
```diff
@@ -7,27 +7,34 @@ Kafka-based storage for Zipkin.
...
+( collected-spans )-|->[ span-consumer ]        [ aggregation ] [ span-store ]--+-->( traces       )
```
Comment (adriancole) [inline README.md]: more seriously than last comment.. heh, there are two major class of filter/mutators span and trace based. the former being stateless so can apply before your aggregations and possibly lower the total size. will start on that today based on this thread https://github.com/openzipkin/zipkin/pull/2072

File: autoconfigure/src/main/resources/zipkin-server-kafkastore.yml (hunk 1 above)
Comment (jeqo) [reply autoconfigure/src/main/resources/zipkin-server-kafkastore.yml]: version 1 of kafka storage topics, not related to zipkin api versions at least. I'm just used to version topics :)

File: README.md (hunk 3 above)
Comment (jeqo) [reply README.md]: Awesome!! 

I found the combination of Kafka Streams [Session Windows](https://kafka.apache.org/20/documentation/streams/developer-guide/dsl-api.html#session-windows) and [Supression](https://www.confluent.io/blog/kafka-streams-take-on-watermarks-and-triggers) as a great fit to enable trace-based filters. e.g. It helped to finally get accurate dependency numbers.

**Pull request merged**